sys/compat/linprocfs/linprocfs.c

```diff
@@ -2092,6 +2095,158 @@ linprocfs_domax_map_cnt(PFS_FILL_ARGS)
 	return (0);
 }
 
+/*
+ * Filler function for proc/sysvipc/msg
+ */
+static int
+linprocfs_dosysvipc_msg(PFS_FILL_ARGS)
+{
+	struct msqid_kernel *msqids;
+	u_long id, msgmni;
+	size_t sz;
+	int error;
+
+	sbuf_printf(sb,
+	    "%10s %10s %4s  %10s %10s %5s %5s %5s %5s %5s %5s %10s %10s %10s\n",
+	    "key", "msqid", "perms", "cbytes", "qnum", "lspid", "lrpid",
+	    "uid", "gid", "cuid", "cgid", "stime", "rtime", "ctime");
+
+	sz = sizeof(struct msqid_kernel) * msginfo.msgmni;
+	msqids = malloc(sz, M_TEMP, M_WAITOK);
+	if (msqids == NULL)
+		return (ENOMEM);
+
+	error = kernel_sysctlbyname(curthread, "kern.ipc.msqids", msqids, &sz,
```

Comment:
This is quite strange way of fetching the data. For instance, if the 32bit process is issuing this sysctl, then the handler adapts and provide quite different 32bit structures instead of what you expect there.

Reply:
> This is quite strange way of fetching the data. For instance, if the 32bit process is issuing this sysctl, then the handler adapts and provide quite different 32bit structures instead of what you expect there.

What do you suggest in this case?

Reply:
I would consider refactoring the sysctls handlers to split them into functions that returns the array of descriptions (perhaps taking the bool compat32 arg or such), and handlers that call that functions. Then the linprocfs code can reuse the array functions.

I did not looked too close how easy such split is, or what other problems could arise.

Reply:
> I would consider refactoring the sysctls handlers to split them into functions that returns the array of descriptions (perhaps taking the bool compat32 arg or such), and handlers that call that functions. Then the linprocfs code can reuse the array functions.
> 
> I did not looked too close how easy such split is, or what other problems could arise.

Too much hassle considering that `WARNING: 32-bit kernels are deprecated and may be removed in FreeBSD 15.0.`.

I installed the 32-bit version and will check what the issue is.

Reply:
This compiled ok on i386.

Reply:
compat32 is not going anywhere.  Problem is not with 32bit kernels, but with 32bit processes running on host (64bit) kernels.  The kernels with compat32 option adapt the result to bitness of the process that queried the sysctl, and the reply is no longer structured in the way you supposed.

Reply:
The problem is not with binaries, but with kernel that internally (and rudely) issues these sysctls and expects that the result is always is in host format.  32bit binaries doing the sysctls are fine because they do want 32bit structures, and kernel code expected that only 32bit binaries do that, not kernel.